python_modules/dagster/dagster/_core/execution/context/metadata_logging.py

```diff
@@ -10,25 +11,74 @@ class OutputMetadataHandle:
     mapping_key: Optional[str]
 
 
+@record
+class AssetMetadataHandle:
+    asset_key: AssetKey
+    partition_key: Optional[str]
+
+
 @record
 class OutputMetadataAccumulator:
-    per_output_metadata: Mapping[OutputMetadataHandle, Mapping[str, Any]]
+    per_output_metadata: Mapping[
+        Union[OutputMetadataHandle, AssetMetadataHandle], Mapping[str, Any]
+    ]
 
     @staticmethod
     def empty() -> "OutputMetadataAccumulator":
         return OutputMetadataAccumulator(per_output_metadata={})
 
-    def get_metadata(self, output_name: str, mapping_key: Optional[str]) -> Mapping[str, Any]:
-        handle = OutputMetadataHandle(output_name=output_name, mapping_key=mapping_key)
+    def get_output_metadata(
+        self, output_name: str, mapping_key: Optional[str]
+    ) -> Mapping[str, Any]:
+        handle = OutputMetadataHandle(
+            output_name=output_name,
+            mapping_key=mapping_key,
+        )
+        return self.per_output_metadata.get(handle, {})
+
+    def get_asset_metadata(
+        self, asset_key: AssetKey, partition_key: Optional[str]
+    ) -> Mapping[str, Any]:
+        handle = AssetMetadataHandle(
+            asset_key=asset_key,
+            partition_key=partition_key,
+        )
         return self.per_output_metadata.get(handle, {})
 
-    def with_additional_metadata(
-        self, output_name: str, mapping_key: Optional[str], metadata: Mapping[str, Any]
+    def with_additional_output_metadata(
+        self,
+        output_name: str,
+        mapping_key: Optional[str],
+        metadata: Mapping[str, Any],
+    ) -> "OutputMetadataAccumulator":
+        return self.new_with_handled_metadata(
+            handle=OutputMetadataHandle(
+                output_name=output_name,
+                mapping_key=mapping_key,
+            ),
+            metadata=metadata,
+        )
+
+    def new_with_handled_metadata(
```

Comment:
placeholder name? maybe just `with_handled_metadata`? also "handled metadata is slightly odd" (parses as "metadata that has been handled" rather than "metadata associated with a handled object"), so maybe just `def _with_metadata()` (private seems reasonable in this case)